analysis/dataflow/report.go

```diff
@@ -106,3 +109,108 @@ func (s *AnalyzerState) ReportSummaryNotConstructed(callSite *CallNode) {
+	UnsafeUsages map[token.Position]string
+	// ReflectUsages records the locations where 'reflect' is used, with a short explanation.
+	ReflectUsages map[token.Position]string
+}
+
+// reportUnsoundFeatures logs warning messages when unsound features are used in a function. Those include:
+//
+// - call to recover builtin
+//
+// - unsafe
+//
+// - reflect
+//
+// Usages of those features are logged at WARN level with the position of where the feature is used.
+func reportUnsoundFeatures(state *AnalyzerState, f *ssa.Function) {
+	unsoundFeatures := FindUnsoundFeatures(f)
+	if len(unsoundFeatures.Recovers) > 0 ||
+		len(unsoundFeatures.UnsafeUsages) > 0 ||
+		len(unsoundFeatures.ReflectUsages) > 0 {
+		msg := fmt.Sprintf("Function %s is using features that may make the analysis unsound.\n", f.String())
+
+		if len(unsoundFeatures.Recovers) > 0 {
+			msg += "    Using recover at position:\n"
+		}
+		for pos := range unsoundFeatures.Recovers {
+			msg += "\t  " + pos.String() + "\n"
+		}
+
+		if len(unsoundFeatures.UnsafeUsages) > 0 {
+			msg += "    Usages of unsafe:\n"
+		}
+
+		for pos, usageMsg := range unsoundFeatures.UnsafeUsages {
+			msg += "      At " + pos.String() + " : " + usageMsg + "\n"
+		}
+
+		if len(unsoundFeatures.ReflectUsages) > 0 {
+			msg += "    Usages of reflection:\n"
+		}
+
+		for pos, usageMsg := range unsoundFeatures.ReflectUsages {
+			msg += "      At " + pos.String() + " : " + usageMsg + "\n"
+		}
+		msg += "    Adding a predefined summary might help avoid soundness issues.\n"
+
+		state.Logger.Warnf(msg)
+	}
+}
+
+// FindUnsoundFeatures returns a record of the unsound features used by a function, if any.
+func FindUnsoundFeatures(f *ssa.Function) UnsoundFeaturesMap {
+	unsafeUsages := map[token.Position]string{}
+	recovers := map[token.Position]bool{}
+	reflectUsages := map[token.Position]string{}
+	lang.IterateInstructions(f, func(index int, instrI ssa.Instruction) {
+		iPos := instrI.Parent().Prog.Fset.Position(instrI.Pos())
+		switch instr := instrI.(type) {
+		case ssa.CallInstruction:
+			callCommon := instr.Common()
+			if callCommon.Value == nil {
+				return
+			}
+			if callCommon.Value.Name() == "recover" {
+				recovers[iPos] = true
+				return
+			}
+			if callCommon.IsInvoke() {
+				// Only warn for implementations.
+				return
+			}
+			typStr := callCommon.Value.Type().String()
+			if strings.Contains(typStr, "unsafe") {
+				unsafeUsages[iPos] = fmt.Sprintf("Calling %s from unsafe package.", callCommon.Value.Name())
+				return
+			}
+			if strings.Contains(typStr, "reflect") {
+				reflectUsages[iPos] = fmt.Sprintf("Calling %s from reflect package.", callCommon.Value.Name())
+				return
+			}
+
+		case *ssa.Alloc:
+			typ := instr.Type().Underlying()
+			if strings.HasPrefix(typ.String(), "unsafe") {
+				unsafeUsages[iPos] = fmt.Sprintf("Allocating object of type %s", typ.String())
+				return
+			}
+			if strings.HasPrefix(typ.String(), "reflect") {
```

Comment:
nit: This `strings.HasPrefix` check is different from the `strings.Contains` check.

I'm also wondering if there's a more standard way to get the name of a package? Maybe via `go/packages.Package.PkgPath`? https://pkg.go.dev/golang.org/x/tools/go/packages#Package

Reply:
Good point, we actually already has the functionality in the `lang` package.